Framework/DataHandling/src/LoadILLDiffraction.cpp

```diff
@@ -193,8 +204,167 @@ void LoadILLDiffraction::loadMetaData() {
+  scanningWorkspaceBuilder.setRelativeRotationsForScans(
+      std::move(tubeAngles), rotationCentre, rotationAxis);
+
+  m_outWorkspace = scanningWorkspaceBuilder.buildWorkspace();
+}
+
+/**
+ * Get the position of the component in the workspace which corresponds to the
+ *angle stored in the scanned variables of the NeXus files. For 1D detectors
+ *this should be the first detector (ID 1), while for 2D detectors (D2B only) it
+ *should be the position of the first tube.
+ *
+ * @param instrumentWorkspace The empty workspace containing the instrument
+ * @return A V3D object containing the position of the relevant component
+ */
+V3D LoadILLDiffraction::getReferenceComponentPosition(
+    const MatrixWorkspace_sptr &instrumentWorkspace) {
+  if (m_instName == "D2B") {
+    return instrumentWorkspace->getInstrument()
+        ->getComponentByName("tube_1")
+        ->getPos();
+  }
+
+  const auto &detInfo = instrumentWorkspace->detectorInfo();
+  const auto &indexOfFirstDet = detInfo.indexOf(1);
+  return detInfo.position(indexOfFirstDet);
+}
+
+/**
+ * Convert from absolute rotation angle, around the sample, of tube 1, to a
+ *relative rotation angle around the sample.
+ *
+ * @param tubeRotations Input is the absolute rotations around the sample of
+ *tube 1, output is the relative rotations required from the IDF for all
+ *detectors
+ * @param firstTubePosition A V3D object containing the position of the first
+ *tube
+ */
+void LoadILLDiffraction::calculateRelativeRotations(
+    std::vector<double> &tubeRotations, const V3D &firstTubePosition) {
+  // The rotations in the NeXus file are the absolute rotation of the first
+  // tube. Here we get the angle of that tube as defined in the IDF.
+
+  double firstTubeRotationAngle =
+      firstTubePosition.angle(V3D(0, 0, 1)) * rad2deg;
+
+  if (m_instName == "D20")
+    firstTubeRotationAngle += D20_NUMBER_DEAD_PIXELS * D20_PIXEL_SIZE;
+
+  g_log.debug() << "First tube rotation:" << firstTubeRotationAngle << "\n";
+
+  // Now pass calculate the rotations to apply for each time index.
+  std::transform(
+      tubeRotations.begin(), tubeRotations.end(), tubeRotations.begin(),
+      [&](double angle) { return (angle - firstTubeRotationAngle); });
+
+  g_log.debug() << "Instrument rotations to be applied : "
+                << tubeRotations.front() << " to " << tubeRotations.back()
+                << "\n";
+}
+
+/**
+ * Fills the counts for the instrument with moving detectors.
+ *
+ * @param data : detector data
+ * @param scan : scan data
+ */
+void LoadILLDiffraction::fillMovingInstrumentScan(const NXUInt &data,
+                                                  const NXDouble &scan) {
+
+  std::vector<double> axis = {0.};
+  std::vector<double> monitor = getMonitor(scan);
+
+  // First load the monitors
+  for (size_t i = 0; i < NUMBER_MONITORS; ++i) {
+    for (size_t j = 0; j < m_numberScanPoints; ++j) {
+      m_outWorkspace->mutableY(j + i * m_numberScanPoints) = monitor[j];
```

Comment:
maybe one should name a temporary for `j + i *m_...` , the same for static instrument workspace filling?